### PR TITLE
bitfields2

### PR DESCRIPTION
**Pull request by marler8997**
Based on discussion with @adamdruppe (http://forum.dlang.org/thread/ibwazxrngcsrfolgbefs@forum.dlang.org)

Added a new version of `bitfields` called `bitfields2`.  The reason for creating `bitfields2` instead of enhancing `bitfields` is because the design requires it to be a mixin template instead of a normal template.  This change is crucial to how `bitfields2` works, but means that any application using it must omit the extra set of parenthesis that `bitfields` requires, i.e.
```D
mixin(bitfields!(int, "a", 8));
mixin bitfields2!(int "a", 8);
```

The main advantage of `bitfields2` is that it is able to reference the field types by alias, whereas the current implementation converts each field type alias to a string and then mixes in the type name.  I've included a small program to demonstrate the differences in each technique and where the weakness lies in the current technique.

```D
import std.stdio, std.typecons;

// Each template allows you to declare a variable
template declareUsingCurrentBitfieldsTechnique(T...)
{
    enum declareUsingCurrentBitfieldsTechnique = T[0].stringof ~ " " ~ T[1] ~ ";";
}
mixin template declareUsingNewBitfieldsTechnique(T...)
{
    mixin("T[0] " ~ T[1] ~ ";");
}

void main()
{
    // example on how to use both, declare varables a and b using each technique
    {
        mixin(declareUsingCurrentBitfieldsTechnique!(int, "a"));
        mixin declareUsingNewBitfieldsTechnique!(int, "b");
        
        // use the declared variables
        a = 3;
        b = 8;
        writefln("a = %s, b = %s", a, b);
    }
    
    // show problem with current technique
    {
        // The current technique fails to compile because wen you convert T[0] to a string,
        // it refers to the template name "Flag" not the template instantiation "Flag!"myflag"".
        // There are ways to try to get the string for the template instantiation but they are
        // brittle and a general solution is very difficult and may not be possible with the current language features.
        mixin(declareUsingCurrentBitfieldsTechnique!(Flag!"myflag", "x"));
        mixin declareUsingNewBitfieldsTechnique!(Flag!"myflag", "y");
    }
}
```

P.S. I've also added the `bitfields2Code` template so that an application can see/grok the generated bitfield code if needed, i.e.
```D
pragma(msg, bitfields2Code!(int, "a", 8));
```

### CODE REVIEW

**Comment by UplinkCoder** [comment]
did you benchmark the compiletime of this, against bitfields ?

**Comment by marler8997** [comment]
The new one was compiling a bit slower because I was using std.format to generate code.  I've since replaced it with string appending, like the original one uses and now performance is basically the same.  Here's the benchmark.

```D
import std.stdio, std.conv, std.bitmanip;

void main()
{
    mixin(function()
    {
        string code;
        foreach(i; 0..300)
        {
            version(Original)
            {
                code ~= 
                    `static struct Bitfield`~ i.to!string ~ `
                    {
                        mixin(bitfields!(
                            uint, "x",    2,
                            int,  "y",    3,
                            uint, "z",    2,
                            bool, "flag", 1));
                    }`;
            }
            version(New)
            {
                code ~= 
                    `static struct Bitfield`~ i.to!string ~ `
                    {
                        mixin bitfields2!(
                            uint, "x",    2,
                            int,  "y",    3,
                            uint, "z",    2,
                            bool, "flag", 1);
                    }`;
            }
        }
        return code;
    }());
}
````

**Comment by marler8997** [comment]
@UplinkCoder 
Update, I found a place in the code where I could combine a string concatenation, and after doing that the new one seems to consistently build faster.  Very odd that string concatenation at compile time can make such a difference.

**Comment by UplinkCoder** [comment]
yeah `std.format` is a compile-time eating beast.

File: std/bitmanip.d [hunk 1]
```diff
@@ -292,7 +272,208 @@ bool), followed by unsigned types, followed by signed types.
...
+}
+private template bitfields2CodeImpl(bool useTypeAlias, T...)
+{
+    enum bitfields2CodeImpl = bitfieldsCodeGenerator();
+
+    // NOTE: the long-verbose name for this function is helpful for compiler error messages
+    private string bitfieldsCodeGenerator()
+    {
+        string store;
+        size_t storageSize = 0;
```
Comment (BBasile) [inline std/bitmanip.d]: Stop creating AST nodes for something that's done automatically (i.e default initialization).

File: std/bitmanip.d [hunk 2]
```diff
@@ -292,7 +272,208 @@ bool), followed by unsigned types, followed by signed types.
...
+        else if (storageSize == uint.sizeof * 8)
+            storageType = "uint";
+        else if (storageSize == ulong.sizeof * 8)
+            storageType = "ulong";
+        else
+            assert(0, "bitfield widths sum to " ~ myToString(storageSize) ~ " but must sum to 8, 16, 32, or 64");
+
+        string mixinCode = "private " ~ storageType ~ " " ~ store ~ ";\n";
+
+        size_t bitOffset = 0;
```
Comment (BBasile) [inline std/bitmanip.d]: ditto

**Comment by adamdruppe** [comment]
On Thu, Jun 22, 2017 at 06:05:04PM -0700, Basile Burg wrote:
> Stop creating AST nodes for something that's done automatically (i.e default initialization).

why?

File: std/bitmanip.d (hunk 1 above)
Comment (CyberShadow) [inline std/bitmanip.d]: https://github.com/dlang/phobos/pull/5497#discussion_r123669334

File: std/bitmanip.d (hunk 2 above)
Comment (CyberShadow) [inline std/bitmanip.d]: https://github.com/dlang/phobos/pull/5497#discussion_r123669334

**Comment by marler8997** [comment]
Seems there is a problem with circleci.sh...is this a known error?  If it's been fixed can someone retrigger the build?

```
./circleci.sh has_public_example
+ HOST_DMD_VER=2.068.2
+ DSCANNER_DMD_VER=2.071.2
++ curl --version
++ head -n 1
+ CURL_USER_AGENT='CirleCI curl 7.22.0 (x86_64-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3'
+ DUB=/home/ubuntu/dlang/dub/dub
+ N=2
+ CIRCLE_NODE_INDEX=0
+ case $CIRCLE_NODE_INDEX in
+ MODEL=64
+ case $1 in
+ echo 'Unknown command'
Unknown command
+ exit 1

./circleci.sh has_public_example returned exit code 1
```

**Comment by wilzbach** [comment]
> Seems there is a problem with circleci.sh...is this a known error?

Yes.

> If it's been fixed can someone retrigger the build?

You need to rebase to master..

**Comment by andralex** [comment]
@marler8997 can you please show a simple example "this would not work (or would be awful) with `bitfields`, but would work very nicely with `bitfields2`"? Thanks!

**Comment by marler8997** [comment]
```D
struct Foo
{
    mixin(bitfields!(
        Flag!"foo", "foo", 1, // Error: std.typecons.Flag(string name) is used as a type
        void, null, 7));
}
```
```D
struct Foo
{
    mixin bitfields2!(
        Flag!"foo", "foo", 1,
        void, null, 7);
}
```

**Comment by marler8997** [comment]
I should also mention that I created another PR to fix the specific use case I just gave as an example, but it is brittle and can't work in general, hence, why this PR was created which does work in the general case.

**Comment by wilzbach** [comment]
I don't want to start a bikeshedding discussion, but `bitfield2` is a rather non-descriptive name (unfortunately I don't have a better suggestion).
Also if we really want to do this, we should at least deprecate `bitfield`, s.t. we don't end up with both for eternity.

File: std/bitmanip.d [hunk 3]
```diff
@@ -9,6 +9,8 @@ $(TR $(TH Category) $(TH Functions))
 $(TR $(TD Bit constructs) $(TD
     $(LREF BitArray)
     $(LREF bitfields)
+    $(LREF bitfields2)
+    $(LREF bitfields2Code)
```
[the remark this answers is not in the document]
Comment (marler8997) [reply std/bitmanip.d]: The purpose of having it isn't to allow a developer to "debug" the implementation but to see/understand what code is being generated.  Maybe that's not a good enough reason though?

[the remark this answers is not in the document]
Comment (marler8997) [reply std/bitmanip.d]: I see in my description that I explicitly said the purpose of this function was to "debug" the code :) Woops!

**Comment by marler8997** [comment]
Cleaning up old PRs that aren't going anywhere

**Pull request closed without merging**